// @mui-treasury/layout/esm/builders/LayoutBuilder.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import HeaderBuilder from './Header';
import EdgeSidebarBuilder from './EdgeSidebar';
import ContentBuilder from './Content';
import FooterBuilder from './Footer';
import InsetSidebarBuilder from './InsetSidebar';
import { DEFAULT_CONTENT_ID, DEFAULT_FOOTER_ID } from '../utils';
export default (function () {
  var initialLayout = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var global = initialLayout.global || {
    autoCollapse: {}
  };
  var header = HeaderBuilder(initialLayout.header);
  var edgeSidebar = EdgeSidebarBuilder(initialLayout.edgeSidebar);
  var insetSidebar = InsetSidebarBuilder(initialLayout.insetSidebar);
  var content = ContentBuilder();
  var footer = FooterBuilder(); // setup default id

  content.create(DEFAULT_CONTENT_ID);
  footer.create(DEFAULT_FOOTER_ID);
  return {
    configureHeader: function configureHeader(callback) {
      callback(header);
    },
    configureEdgeSidebar: function configureEdgeSidebar(callback) {
      callback(edgeSidebar);
    },
    configureInsetSidebar: function configureInsetSidebar(callback) {
      callback(insetSidebar);
    },
    enableAutoCollapse: function enableAutoCollapse(sidebarId) {
      var breakpoint = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'md';
      global.autoCollapse[sidebarId] = breakpoint;
    },
    getComponentData: function getComponentData() {
      return {
        global: global,
        edgeSidebar: edgeSidebar.getData(),
        insetSidebar: insetSidebar.getData(),
        header: header.getData(),
        headerId: header.getId(),
        content: content.getData(),
        footer: footer.getData()
      };
    },
    clone: function clone() {
      return (// use this approach as deep clone for now
        JSON.parse(JSON.stringify({
          header: header.getData(),
          edgeSidebar: edgeSidebar.getData().configMapById,
          insetSidebar: insetSidebar.getData().configMapById,
          content: content.getData(),
          footer: footer.getData()
        }))
      );
    },
    getInitialState: function getInitialState() {
      var ids = edgeSidebar.getSidebarIds();
      return {
        sidebar: ids.reduce(function (result, curr) {
          return _extends(_extends({}, result), {}, _defineProperty({}, curr, {
            open: false,
            collapsed: false
          }));
        }, {})
      };
    },
    debug: function debug() {
      if (process.env.NODE_ENV !== 'production') {
        header.debug();
        edgeSidebar.debug();
        insetSidebar.debug();
        content.debug();
        footer.debug();
      }
    },
    getJSON: function getJSON() {
      console.log(JSON.stringify({
        header: header.getData(),
        edgeSidebar: edgeSidebar.getData().configMapById,
        insetSidebar: insetSidebar.getData().configMapById
      }, null, 2));
    }
  };
});
