// @mui-treasury/layout/esm/builders/EdgeSidebar/EdgeSidebarBuilder.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import { keys } from '@material-ui/core/styles/createBreakpoints';
import { normalizeMapById, attachHiddenToMapById } from '../../utils';
export var isUniqueSidebars = function isUniqueSidebars(sidebars) {
  var keys = [];
  var isUnique = true;
  sidebars.forEach(function (_ref) {
    var id = _ref.id;

    if (!id) {
      throw new Error('[Layout] - All Sidebar must have id');
    }

    if (keys.includes(id)) {
      isUnique = false;
    } else {
      keys.push(id);
    }
  });
  return isUnique;
};
export default (function () {
  var initialMapById = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var hiddenById = {};
  var sidebarIds = Object.keys(initialMapById);

  var mapById = _extends({}, initialMapById);

  var addConfig = function addConfig(breakpoint, config) {
    if (!sidebarIds.includes(config.id)) {
      sidebarIds.push(config.id);
    }

    if (!mapById[config.id]) {
      mapById[config.id] = {};
    }

    mapById[config.id][breakpoint] = config;
  };

  return {
    create: function create(id, props) {
      hiddenById[id] = [];

      var Registry = function Registry() {
        return {
          registerPersistentConfig: function registerPersistentConfig(breakpoint, config) {
            addConfig(breakpoint, _extends(_extends(_extends({}, config), props), {}, {
              id: id,
              variant: 'persistent'
            }));
            return this;
          },
          registerPermanentConfig: function registerPermanentConfig(breakpoint, config) {
            addConfig(breakpoint, _extends(_extends(_extends({}, config), props), {}, {
              id: id,
              variant: 'permanent'
            }));
            return this;
          },
          registerTemporaryConfig: function registerTemporaryConfig(breakpoint, config) {
            addConfig(breakpoint, _extends(_extends(_extends({}, config), props), {}, {
              id: id,
              variant: 'temporary'
            }));
            return this;
          }
        };
      };

      return Registry();
    },
    update: function update(id, updater) {
      if (mapById[id]) {
        updater(mapById[id]);
      } else {
        console.warn("No sidebar to update. id: ".concat(id));
      }
    },
    hide: function hide(id, breakpoints) {
      if (typeof breakpoints === 'boolean') {
        hiddenById[id] = breakpoints ? keys : [];
      } else {
        if (!hiddenById[id]) hiddenById[id] = [];
        hiddenById[id] = breakpoints;
      }
    },
    getData: function getData() {
      var attachedMapById = attachHiddenToMapById(mapById, hiddenById);
      return {
        configMap: normalizeMapById(attachedMapById),
        configMapById: attachedMapById,
        sidebarIds: sidebarIds,
        hiddenById: hiddenById
      };
    },
    getSidebarIds: function getSidebarIds() {
      return sidebarIds;
    },
    debug: function debug() {
      if (process.env.NODE_ENV !== 'production') {
        sidebarIds.forEach(function (id) {
          console.group('EdgeSidebar:', "\"".concat(id, "\""));
          console.table(mapById[id]);
          console.groupEnd();
        });
      }
    }
  };
});
