// @mui-treasury/layout/builders/Header/HeaderBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _utils = require("../../utils");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = function _default() {
  var initialMap = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var id = _utils.DEFAULT_HEADER_ID;

  var map = _extends({}, initialMap);

  var Registry = function Registry(headerId) {
    return {
      registerConfig: function registerConfig(breakpoint, config) {
        map[breakpoint] = _extends(_extends({
          initialHeight: (0, _utils.pickNearestBreakpoint)(_utils.INITIAL_HEADER_HEIGHT, breakpoint)
        }, config), {}, {
          id: headerId
        });
        return this;
      }
    };
  };

  var defaultRegistry = Registry(id);
  return {
    create: function create(headerId) {
      id = headerId;
      map = _extends({}, initialMap); // reset map

      return Registry(headerId); // return new registry
    },
    registerConfig: defaultRegistry.registerConfig,
    update: function update(updater) {
      updater(map);
    },
    getId: function getId() {
      return id;
    },
    getData: function getData() {
      return map;
    },
    debug: function debug() {
      if (process.env.NODE_ENV !== 'production') {
        console.group('Header:', "\"".concat(id, "\""));
        console.table(map);
        console.groupEnd();
      }
    }
  };
};

exports["default"] = _default;
